### PR TITLE
Test results of some universal functions

### PR DESCRIPTION
**Pull request by joaosferreira**
I'm adding some more tests as [suggested](https://github.com/Quansight-Labs/udiff/pull/5#issuecomment-588768318).

### CODE REVIEW

File: unumpy/tests/test_numpy.py [hunk 1]
```diff
@@ -281,3 +281,40 @@ def test_array_creation(backend, method, args, kwargs):
...
+        (np.true_divide, ([6, 1], [3, 2]), {}, [2.0, 0.5]),
+        (np.power, ([2, 3], [3, 2]), {}, [8, 9]),
+        (np.positive, ([1, -2],), {}, [1, -2]),
+        (np.negative, ([-2, 3],), {}, [2, -3]),
+        (np.conj, ([1.0 + 2.0j, -1.0 - 1j],), {}, [1.0 - 2.0j, -1.0 + 1j]),
+        (np.exp, ([0, 1, 2],), {}, [1.0, 2.718281828459045, 7.38905609893065]),
+        (np.exp2, ([3, 4],), {}, [8, 16]),
+        (np.log, ([1, np.e, np.e ** 2],), {}, [0.0, 1.0, 2.0]),
+        (np.log2, ([1, 2, 2 ** 4],), {}, [0.0, 1.0, 4.0]),
+        (np.log10, ([1e-5, -3.0],), {}, [-5.0, onp.NaN]),
```
Comment (hameerabbasi) [inline unumpy/tests/test_numpy.py]: This should be `np.nan`, with that being defined as `float('nan')`

**Review by hameerabbasi — COMMENTED** [review]
The problems here are manifold:

* `xnd` doesn't have `reciprocal`
* `numpy.array_equal` doesn't handle Sparse arrays.
* `numpy.array_equal` doesn't compare `nan`s properly.

**Comment by hameerabbasi** [comment]
`onp.equal` won't do either, because according to [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754), `NaN` is never equal to anything, even if it is `NaN`. To keep the code generic, I'd suggest the following:

```python
np.all(np.equal(ret, res) | (np.isnan(ret) & np.isnan(res)))
```

**Comment by hameerabbasi** [comment]
You can add an exception in the `EXCEPTIONS` list for `reciprocal`, similar to the others.

**Comment by joaosferreira** [comment]
> `onp.equal` won't do either, because according to [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754), `NaN` is never equal to anything, even if it is `NaN`.

I forgot about this, thanks!

> ```python
> np.all(np.equal(ret, res) | (np.isnan(ret) & np.isnan(res)))
> ```

I think this should be logical AND instead of OR, right?

**Comment by hameerabbasi** [comment]
No. Either they should be equal, OR both of them should be `NaN`. And the `np.all` is to collapse the boolean array into a single value.

**Comment by joaosferreira** [comment]
Yes, you are right @hameerabbasi.

File: unumpy/tests/test_numpy.py [hunk 2]
```diff
@@ -69,6 +69,7 @@
     (DaskBackend, np.sort_complex),
     (DaskBackend, np.msort),
     (DaskBackend, np.searchsorted),
+    (SparseBackend, np.reciprocal),
```
Comment (hameerabbasi) [inline unumpy/tests/test_numpy.py]: Should be `XndBackend`.

File: unumpy/tests/test_numpy.py [hunk 3]
```diff
@@ -318,6 +319,6 @@ def test_ufuncs_results(backend, method, args, kwargs, res):
         pytest.xfail(reason="The backend has no implementation for this ufunc.")
 
     if isinstance(ret, sparse.SparseArray):
```
Comment (hameerabbasi) [inline unumpy/tests/test_numpy.py]: Don't need this line or the one below.

File: unumpy/tests/test_numpy.py [hunk 4]
```diff
@@ -318,6 +319,6 @@ def test_ufuncs_results(backend, method, args, kwargs, res):
...
-        assert onp.equal(ret.todense(), res.todense())
+        ret, res = ret.todense(), res.todense()
 
-    assert onp.equal(ret, res)
+    assert np.all(np.equal(ret, res) or (np.isnan(ret) and np.isnan(res)))
```
Comment (hameerabbasi) [inline unumpy/tests/test_numpy.py]: There's a reason I used `|` and `&`: They work on arrays element-wise. `or` and `and` attempt to make the input a Python `bool` before proceeding.

Comment (hameerabbasi) [reply unumpy/tests/test_numpy.py]: Which won't succeed because only 0D arrays collapse to a `bool`.

**Review by hameerabbasi — DISMISSED** [review]
A few more comments.

File: unumpy/tests/test_numpy.py (hunk 3 above)
Comment (joaosferreira) [reply unumpy/tests/test_numpy.py]: The `sparse.todense` method seems necessary at least for the `ret` variable, otherwise `np.equal` fails.

**Comment by joaosferreira** [comment]
Only Dask backend seems to be failing on my end now.

**Comment by hameerabbasi** [comment]
Calling in some experts here, because locally, `np.cbrt(27) == 3.0`, but on CI, that doesn't seem to be the case. @peterbell10, can you test this on Linux?

**Comment by grlee77** [comment]
I tried it locally on linux (using NumPy 1.18.1 with OpenBLAS-based NumPy from conda-forge)

```Python
 np.cbrt(27)
```
and got 3.0000000000000004

so `np.cbrt(27) == 3` is False

**edit**: To clarify, the above is via `import numpy as np`, i.e. not going through uarray.

File: unumpy/tests/test_numpy.py [hunk 5]
```diff
@@ -281,3 +282,46 @@ def test_array_creation(backend, method, args, kwargs):
...
+    try:
+        with ua.set_backend(backend, coerce=True):
+            ret = method(*args, **kwargs)
+    except ua.BackendNotImplementedError:
+        if backend in FULLY_TESTED_BACKENDS and backend is not XndBackend:
+            raise
+        pytest.xfail(reason="The backend has no implementation for this ufunc.")
+
+    res = np.asarray(res)
+    assert onp.all(onp.equal(ret, res) | (onp.isnan(ret) & onp.isnan(res)))
```
Comment (peterbell10) [inline unumpy/tests/test_numpy.py]: I would generally avoid using `equal` with floating point values. Here, I would use [assert_allclose(ret, res)](https://docs.scipy.org/doc/numpy/reference/generated/numpy.testing.assert_allclose.html) instead.

Comment (hameerabbasi) [reply unumpy/tests/test_numpy.py]: I'm avoiding using anything specific to NumPy here. Looks like we'll have to wait until [`np.isclose`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.isclose.html) and [`np.allclose`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.allclose.html) are in.

Comment (hameerabbasi) [reply unumpy/tests/test_numpy.py]: @prasunanand Can you possibly add these two functions with defaults?

Comment (joaosferreira) [reply unumpy/tests/test_numpy.py]: I could take the PR to add these two functions if you want @hameerabbasi, @prasunanand.

Comment (hameerabbasi) [reply unumpy/tests/test_numpy.py]: @joaosferreira That'd be nice if you could manage it. 😄

**Comment by hameerabbasi** [comment]
@joaosferreira If you merge with `master`, I added `isclose` and `allclose`.

File: unumpy/tests/test_numpy.py (hunk 5 above)
Comment (hameerabbasi) [reply unumpy/tests/test_numpy.py]: This should be changed to `assert np.allclose(ret, res, equal_nan=True)`.

**Comment by joaosferreira** [comment]
Thank you for your help @hameerabbasi.

File: unumpy/tests/test_numpy.py [hunk 6]
```diff
@@ -285,3 +286,46 @@ def test_array_creation(backend, method, args, kwargs):
...
+    try:
+        with ua.set_backend(backend, coerce=True):
+            ret = method(*args, **kwargs)
+    except ua.BackendNotImplementedError:
+        if backend in FULLY_TESTED_BACKENDS and backend is not XndBackend:
+            raise
+        pytest.xfail(reason="The backend has no implementation for this ufunc.")
+
+    res = np.asarray(res)
+    assert np.allclose(ret, res, equal_nan=True)
```
Comment (hameerabbasi) [inline unumpy/tests/test_numpy.py]: These two lines should be moved inside the `with` clause.

**Review by hameerabbasi — APPROVED** [review]

**Pull request merged**